client/clientutil/modelinfo.go

```diff
@@ -0,0 +1,668 @@
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package clientutil
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+	"text/tabwriter"
+	"time"
+	"unicode"
+
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/strutil"
+	"github.com/snapcore/snapd/timeutil"
+)
+
+var (
+	invalidTypeMessage = i18n.G("invalid type for %q header")
+	errNoMainAssertion = errors.New(i18n.G("device not ready yet (no assertions found)"))
+	errNoSerial        = errors.New(i18n.G("device not registered yet (no serial assertion found)"))
+
+	// this list is a "nice" "human" "readable" "ordering" of headers to print
+	// off, sorted in lexical order with meta headers and primary key headers
+	// removed, and big nasty keys such as device-key-sha3-384 and
+	// device-key at the bottom
+	// it also contains both serial and model assertion headers, but we
+	// follow the same code path for both assertion types and some of the
+	// headers are shared between the two, so it still works out correctly
+	niceOrdering = [...]string{
+		"architecture",
+		"base",
+		"classic",
+		"display-name",
+		"gadget",
+		"kernel",
+		"revision",
+		"store",
+		"system-user-authority",
+		"timestamp",
+		"required-snaps", // for uc16 and uc18 models
+		"snaps",          // for uc20 models
+		"device-key-sha3-384",
+		"device-key",
+	}
+)
+
+type OutputFormat int
+
+type ModelFormatter interface {
+	GetPublisher() string
+	GetEscapedDash() string
+}
+
+const (
+	MODELWRITER_RAW_FORMAT OutputFormat = iota
+	MODELWRITER_YAML_FORMAT
+	MODELWRITER_JSON_FORMAT
+)
+
+// modelAssertJSON is a helper to write out a single assertion in json
+// format that matches how it's done in api_model.go. We have stolen
+// the structure from there.
+type modelAssertJSON struct {
+	Headers map[string]interface{} `json:"headers,omitempty"`
+	Body    string                 `json:"body,omitempty"`
+}
+
+// modelWriterState describes a state of the modelWriter. We use this to support
+// nested objects and arrays.
+type modelWriterState struct {
+	indent       int
+	firstElement bool
+	inArray      bool
+	inlineArray  bool
+	arrayMembers []string
+}
+
+// modelWriter supports building simple yaml and json output for a model. It's
+// built in a 'write-as-you-go' fashion, to simplify how we write members while
+// parsing the model assertions.
+type modelWriter struct {
+	w            *tabwriter.Writer
+	format       OutputFormat
+	indentSpaces int
+	currentState modelWriterState
+	states       []modelWriterState
+}
```

Comment:
As Samuele mentioned in the standup it's a bit unusual and not how we generally do it.  For yaml we generally just print "yaml'ish" output, or in some cases have a simple helper struct and use `yaml.Marshal(..)` to format the output (see `ctlcmd/refresh.go`).  This basically looks like custom yaml and json marshaller, I'm not sure we should go that way (on a side note despite `modelWriter` name it seems to be actually a very generic writer, not limited to models).